### PR TITLE
feat: Initial setup and base map/store substream modules

### PR DESCRIPTION
**Pull request by juanmahidalgo**
Initial setup for substreams. 

It includes a few `map` modules for:
- `map_collection_created` => Reads the `ProxyCreate` events from the Collections v1 and v3 factories (and store them in an `store` module so it's traced across the blocks).
- `map_issues` => Reads the `Issue` event out of a `collections_store` array of collection contract addresses and emits an array of `Issue`s.
- `map_add_items` => Read the `AddItem` event when items are added to the collections created by the factories.
- `db_out`: captures the collections created on each block by `map_collection_created` and the items added to them by the `AddItem` event (mapped in `map_add_items`) and that uses the `postgres` sink module to dump it into a db

This also includes an rs abi that has been crafted manually to avoid the tuple limitation of `ethabi` when generating abis with `Tuple`s.

### CODE REVIEW

File: src/rpc.rs [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+}
+
+pub fn get_collection_item_count(collection_address: Vec<u8>) -> Option<BigInt> {
+    substreams::log::info!("get_collection_item_count2 {:?}", collection_address);
+    let func = abi::collections_v2::functions::ItemsCount {};
+    if let Some(items_result) = func.call(collection_address) {
+        return Some(items_result);
+    };
+
+    return None;
```
Comment (guidota) [inline src/rpc.rs]: I think you can avoid the checks and directly return the result
```suggestion
    fun.call(collection_address)
```

File: src/rpc.rs [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+    substreams::scalar::BigInt,
+    substreams::scalar::BigInt,
+    Vec<u8>,
+    std::string::String,
+    std::string::String,
+)> {
+    let items = abi::collections_v2::functions::Items {
+        param0: BigInt::from(item_id),
+    };
+    if let Some(items_result) = items.call(collection_address) {
```
Comment (guidota) [inline src/rpc.rs]: Same here

File: src/utils.rs [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+use substreams::scalar::BigInt;
+
+use crate::pb::dcl;
+
+pub fn get_item_id(contract_address: String, item_id: String) -> std::string::String {
+    let mut id = contract_address;
+    id.push_str(&String::from("-"));
+    id.push_str(&item_id);
+    id
+}
```
Comment (guidota) [inline src/utils.rs]:
```suggestion
pub fn get_item_id(contract_address: String, item_id: String) -> String {
    format!("{}-{}", contract_address, item_id)
}
```

File: build.rs [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+fn main() -> Result<(), anyhow::Error> {
+    // Abigen::new("CollectionV2", "abi/collectionV2.json")?
+    //     .generate()?
+    //     .write_to_file("src/abi/collectionV2.rs")?;
+
+    Abigen::new("CollectionFactoryV3", "abi/CollectionFactoryV3.json")?
+        .generate()?
+        .write_to_file("src/abi/collectionFactoryv3.rs")?;
+
```
Comment (guidota) [inline build.rs:12]: You can add extra steps here to generate the `modified` version of `collectionV2`

**Review by guidota — COMMENTED** [review]
Excellent! Leaving some minor suggestions.

File: src/db.rs [hunk 5]
```diff
@@ -0,0 +1,56 @@
+use crate::dcl_hex;
+use crate::pb::dcl;
+use crate::utils::sanitize_sql_string;
+use substreams::prelude::BigInt;
+use substreams_database_change::pb::database::{table_change, DatabaseChanges};
+
+pub fn transform_collection_database_changes(
+    changes: &mut DatabaseChanges,
+    collections: dcl::Collections,
```
Comment (cyaiox) [inline src/db.rs]: is it required to use the namespace here?

File: src/rpc.rs (hunk 1 above)
Comment (juanmahidalgo) [reply src/rpc.rs]: oohh true, updated!

File: src/db.rs (hunk 5 above)
Comment (juanmahidalgo) [reply src/db.rs]: I could use just `Collections` but I'd need to modify the top level import to 
```
use crate::{dcl_hex, pb::dcl::Collections};
```
so at the end of the day, I think it's quite similar to use any of these 2 alternatives

**Pull request merged**